configuration/filesystem/shared-storage.sh

```diff
@@ -0,0 +1,414 @@
+    # the only value unavailable to non-HA setups
+    echo_error 'DRBD use case unavailable in non-HA setups, exiting.'
+    exit 1
+fi
+
+
+case $SHARED_FS_TYPE in
+
+    drbd | dev )
+            echo_info "Use case \"$SHARED_FS_TYPE\" selected, setting up the device and FS."
+            ;;
+
+    none | export )
+            echo_info "Use case \"$SHARED_FS_TYPE\" selected. No FS setup required, exiting."
+            exit 0
+            ;;
+
+    * )     echo_error 'Invalid shared storage use case, exiting.'
+            exit 1
+
+esac
+
+
+
+################################################################################
+#                                                                              #
+# From that point onward we're only dealing with 'drbd' and 'dev' use cases.   #
+#                                                                              #
+################################################################################
+
+
+#---------------------------------------
+# Use case: "drbd"
+#---------------------------------------
+
+
+if [[ $SHARED_FS_TYPE == drbd ]] ; then
+    
+    # Prepare the device
+
+    check_block_device
+    SHARED_FS_DRBD_DEVICE=/dev/drbd1
+    display_var SHARED_FS_{,DRBD_}DEVICE
+
+    if flag_is_unset SHARED_FS_NO_FORMAT ; then
+        sgdisk -Z ${SHARED_FS_DEVICE} 2>/dev/null
+        # some strange timing issues now and then, so:
+        sleep 2 ; sync
+    fi
+
+    # Set up DRBD and check that it's good
+
+    install_drbd_config
+    flag_is_unset SHARED_FS_NO_FORMAT && create_drbd_metadata
+
+
+    #---------------------------------------
+    # HA primary
+    #---------------------------------------
+
+    if flag_is_set PRIMARY_INSTALL ; then
+
+        # Set up the local resource and filesystem
+
+        start_drbd
+
+        echo_info 'Setting DRBD resource as primary'
+        if ! drbdadm -- --overwrite-data-of-peer primary trinity_disk ; then
+            echo_error 'Failed to set DRBD resource as primary, exiting.'
+            exit 1
+        fi
+
+        # Alright now we can format the thing
+        flag_is_unset SHARED_FS_NO_FORMAT && format_device /dev/drbd/by-res/trinity_disk
+
+        flag_is_unset QUIET && { echo ; cat /proc/drbd ; echo ; }
+
+        # It will be started again later by the pacemaker resource
+        stop_drbd
+
+
+        # Set up the Pacemaker resources
+
+        echo_info 'Setting up the Pacemaker resources'
+
+        tmpfile=$(mktemp -p /root pacemaker_drbd.XXXX)
+        pcs cluster cib $tmpfile
+
+        # The pair of resources for the DRBD service
+        pcs -f $tmpfile resource create DRBD ocf:linbit:drbd drbd_resource=trinity_disk op monitor interval=59s promote interval=67s
+        pcs -f $tmpfile resource master Trinity-drbd DRBD master-max=1 master-node-max=1 clone-max=2 clone-node-max=1 notify=true
+
+        # The DRBD RA returns way too early, we need a short delay
```

Comment:
Did you try to avoid groups, but using order/collocation constraints instead? I found them more reliable.